Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +24,62 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return amphour_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return (current_ < 0);
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return QTime(0, minutes, seconds, miliseconds);
 }
 
+int BatteryStateOfChargeService::getHours() const
+{
+    return hoursInteger;
+}
+
+
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
     // This is where you can update your variables
     // Hint: There are many different ways that the totalAmpHoursUsed can be updated
     // i.e: Taking a running average of your data values, using most recent data points, etc.
+    double currentOld_ = current_;
+    current_ = batteryData.current;
+
+    QTime timeOld_ = time;
+    time = batteryData.time;
+
+    double averageAmphour_ = (current_ + currentOld_) / 2;
+
+    //timeOld_ would be null for the first entry of data
+    if (!timeOld_.isNull())
+    {
+        amphour_ += averageAmphour_ * timeOld_.msecsTo(time) / HOURS_TO_MILISECONDS;
+    }
+
+    //Calculation of the time until charged/depleted
+    if (isCharging())
+    {
+        totalTimeHoursDouble =  abs(amphour_ / current_);
+    }
+    else
+    {
+        totalTimeHoursDouble =  (BATTERY_AMP_HOUR_CAPACITY - amphour_) / current_;
+    }
+
+    //Calculating the Time
+    seconds = totalTimeHoursDouble*HOURS_TO_SECONDS;
```

Comment:
Keep a space between operators. `totalTimeHoursDouble * HOURS_TO_SECONDS`, this goes for the following lines too.